### DISCUSSION TITLE
API interaction with the back-end

### DISCUSSION THREAD

**Discussion by MohammedAlmatrafi** [UniTime/unitime#166] [Q&A] [answered]
Hello, we are working on a project that utilizes UniTime as backend service through API calls,
Our goal is to simplifiy the course timetabling process for our needs and UniTime is a great timetabling solution that we think is gonna help us. 
There are few steps to achieve our goal which can be done through the RESTful API:
1. Using the POST api end point of (/api/exchange) we were able to load all the necessary information to begin the timetable.
2. Run the solver and commit the timetable.
3. Using the data exchange (GET) api again we can export the current timetable in a specific term.

The only thing that's missing is loading the solver, starting the solver, (after some time)stop the solver, save the best timetable, & commit the timetable, and finally unload the solver.

My question is **are we able to do all that using the scripts API?**
if yes, **how to do it?** give us a guide or a hint please.

**Comment by tomas-muller** [comment, marked as answer]
There is an option not to use UniTime at all. It is possible to run the solver by itself -- as a Java program, either called programmatically (from within Java) or from a command line. It takes an XML file (with the course timetabling problem), runs for the given amount of time, and produces a new XML file containing the solution. See [github.com/UniTime/cpsolver](https://github.com/UniTime/cpsolver) for more details. Please note that it takes a different XML (see the format [here](https://www.cpsolver.org/uct_dataformat_v24.php), [which can be exported from UniTime](https://help.unitime.org/exporting-solver-xml)).

Back to your question, it is possible to start/stop/etc the solver using the JSON RPC calls and simulating the [Solver](https://help.unitime.org/course-timetabling-solver) page with the [SolverPageRequest](https://github.com/UniTime/unitime/blob/master/JavaSource/org/unitime/timetable/gwt/shared/SolverInterface.java#L399) and [SolverPageResponse](https://github.com/UniTime/unitime/blob/master/JavaSource/org/unitime/timetable/gwt/shared/SolverInterface.java#L247) messages. Here is an example:

1. To load the page, use operation=INIT
```
curl -uadmin:admin -k -X POST -H "Content-Type:application/json;charset=UTF-8" -d @solver-init.json \
  "http://demo.unitime.org/UniTime/api/json?term=Fal2010woebegon&type=org.unitime.timetable.gwt.shared.SolverInterface%24SolverPageRequest"
```

Where `solver-init.json` contains
```
{
    type: "COURSE",
    operation: "INIT"
}
```
This will return the list of available solver configurations and other parameters shown on the page, together with the owners (solver groups) for which the course timetabling solver can be run. See the [Course Timetabling Solver](https://help.unitime.org/course-timetabling-solver) on our [online demo](https://demo.unitime.org) to interpret the data. It will also show if there is a solver already running for the user (each user can only have one course timetabling solver running at a time).

2. To load and start the solver, you can do the same call with a different message:
```
{
    type: "COURSE",
    operation: "START",
    ownerIds: [365, 366],
    configurationId: 5
}
```
Operation is START (which will also do the load if there is no solver loaded in), than there needs to be the list of solver groups (owners) for which the solver should run and the solver configuration to use. You do not need to set any other parameters as UniTime will take the defaults in this case -- which you can change on the [Solver Configurations](https://help.unitime.org/solver-configurations) page.

In particular, you can change the `When finished` parameter to `Save as New and Unload` for the solver to automatically save the results as a new solution and get unloaded once it is done. You can also change the solver run time in the `Maximal solver time (in sec)` parameter.

**Reply by MohammedAlmatrafi** [reply]
Thank you for your reply and the valuable information.
I appreciate it 🌹
